src/query/subquery_cache.c

```diff
@@ -0,0 +1,449 @@
+
+  xasl->sq_cache->ht = mht_create ("sq_cache", sq_hm_entries, sq_hash_func, sq_cmp_func);
+  if (!xasl->sq_cache->ht)
+    {
+      return ER_FAILED;
+    }
+  xasl->sq_cache->enabled = true;
+  xasl->sq_cache->size_max = max_subquery_cache_size;
+  return NO_ERROR;
+}
+
+/*
+ * sq_put () - Puts a value into the cache for a given XASL node.
+ *   return: NO_ERROR if the value is successfully cached, ER_FAILED otherwise.
+ *   xasl(in): The XASL node for which the value is being cached.
+ *   regu_var(in): The regu variable containing the value to be cached.
+ *
+ * This function attempts to cache the result of a regu variable associated with a given XASL node. It generates a key based on
+ * the XASL node's structure and creates a cache entry if such a key does not already exist in the cache. The function returns
+ * NO_ERROR if the value is successfully cached, or ER_FAILED if the key could not be generated or the value could not be added
+ * to the cache.
+ */
+int
+sq_put (THREAD_ENTRY * thread_p, SQ_KEY * key, XASL_NODE * xasl, REGU_VARIABLE * regu_var)
+{
+  SQ_VAL *val;
+  const void *ret;
+  UINT64 new_entry_size = 0;
+  int i;
+
+  val = sq_make_val (thread_p, regu_var);
+
+  if (!xasl->sq_cache->ht)
+    {
+      sq_cache_initialize (thread_p, xasl);
+    }
+
+  for (i = 0; i < key->n_elements; i++)
+    {
+      new_entry_size += (UINT64) or_db_value_size (key->dbv_array[i]);
+    }
+  new_entry_size += sizeof (SQ_KEY);
+  switch (val->t)
+    {
+    case TYPE_CONSTANT:
+      new_entry_size += (UINT64) or_db_value_size (val->val.dbvalptr) + sizeof (SQ_VAL);
+      break;
+
+    default:
+      break;
+    }
+
+  if (xasl->sq_cache->size_max < xasl->sq_cache->size + new_entry_size)
+    {
+      xasl->sq_cache->enabled = false;
+      sq_free_val (val);
+      return ER_FAILED;
+    }
+
+  ret = mht_put_if_not_exists (xasl->sq_cache->ht, key, val);
+
+  if (!ret || ret != val)
+    {
+      sq_free_val (val);
+      return ER_FAILED;
+    }
+  xasl->sq_cache->size += new_entry_size;
+  return NO_ERROR;
+}
+
+/*
+ * sq_get () - Retrieves a value from the cache for a given XASL node.
+ *   return: True if a cached value is found and retrieved, False otherwise.
+ *   xasl(in): The XASL node for which a cached value is being retrieved.
+ *   regu_var(in/out): The regu variable where the retrieved value will be stored.
+ *
+ * This function attempts to retrieve a value from the cache for a given XASL node. It generates a key based on the XASL node's
+ * structure and looks up the cache for a matching value. If a cached value is found, it is unpacked into the specified regu
+ * variable, and the function returns True. Otherwise, the function updates cache miss counters and returns False.
+ */
+bool
+sq_get (THREAD_ENTRY * thread_p, SQ_KEY * key, XASL_NODE * xasl, REGU_VARIABLE * regu_var)
+{
+  SQ_VAL *ret;
+
+  if (XASL_IS_FLAGED (xasl, XASL_SQ_CACHE) && xasl->sq_cache->ht)
```

Comment:
I think it is better not to check the flag.

Reply:
Okay.